### ISSUE TITLE
Segfaults since v1.15.0

### ISSUE THREAD

**Issue by ngardiner** [adriankumpf/teslamate#486]
Since I upgraded to TeslaMate v1.15.0, I've been seeing regular segfaults that tend to occur during activity (such as a charging session). As I've retained logs, I can see that the instances of Segmentation Fault have been:

- v1.14.3 - 0
- v1.15.0 - 2
- v1.15.1 - 6
- v1.16.0 - 6

It is possible that an erlang/elixir update may be responsible, it is not necessarily the version of TeslaMate however this is the correlation I have so far.

**To Reproduce**

This occurs on an ongoing basis.

**Relevant entries from the logs**

The following shows the crashes I can see in the logs:

```
./start.sh: line 18:  7240 Segmentation fault      _build/prod/rel/teslamate/bin/teslamate start
./start.sh: line 18: 28247 Segmentation fault      _build/prod/rel/teslamate/bin/teslamate start
./start.sh: line 18:  9186 Segmentation fault      _build/prod/rel/teslamate/bin/teslamate start
./start.sh: line 18: 18864 Segmentation fault      _build/prod/rel/teslamate/bin/teslamate start
./start.sh: line 18: 10593 Segmentation fault      _build/prod/rel/teslamate/bin/teslamate start
./start.sh: line 18: 23913 Segmentation fault      _build/prod/rel/teslamate/bin/teslamate start
./start.sh: line 18: 32490 Segmentation fault      _build/prod/rel/teslamate/bin/teslamate start
```

Analyzing the top of the last erl_crash.dump (they overwrite unfortunately), I see:

```
=erl_crash_dump:0.5
Wed Feb 12 12:46:51 2020
Slogan: Kernel pid terminated (application_controller) ({application_start_failure,teslamate,{{shutdown,{failed_to_start_child,'Elixir.TeslaMate.Vehicles',{#{'__exception__' => true,'__struct__' => 'Elixir.Po
System version: Erlang/OTP 22 [erts-10.6.4] [source] [64-bit] [smp:2:1] [ds:2:1:10] [async-threads:1] [hipe]
Compiled: Wed Feb  5 16:20:20 2020
Taints: dyntrace,crypto,asn1rt_nif
Atoms: 43965
Calling Thread: scheduler:1
=scheduler:1
Scheduler Sleep Info Flags:
Scheduler Sleep Info Aux Work: THR_PRGR_LATER_OP
Current Port:
Run Queue Max Length: 0
Run Queue High Length: 0
Run Queue Normal Length: 0
Run Queue Low Length: 0
Run Queue Port Length: 0
Run Queue Flags: OUT_OF_WORK | HALFTIME_OUT_OF_WORK | NONEMPTY | EXEC
Current Process: <0.0.0>
Current Process State: Running
Current Process Internal State: ACT_PRIO_NORMAL | USR_PRIO_NORMAL | PRQ_PRIO_NORMAL | ACTIVE | RUNNING
Current Process Program counter: 0x00007f22ce65ca20 (init:printable_list/1 + 24)
Current Process CP: 0x00007f22ce65d6c8 (init:boot_loop/2 + 1944)
Current Process Limited Stack Trace:
0x00007f228f5d1c08:SReturn addr 0xF6B03DF8 (<terminate process normally>)
=scheduler:2
Scheduler Sleep Info Flags: SLEEPING | TSE_SLEEPING | WAITING | SUSPENDED
Scheduler Sleep Info Aux Work:
Current Port:
Run Queue Max Length: 0
Run Queue High Length: 0
Run Queue Normal Length: 0
Run Queue Low Length: 0
Run Queue Port Length: 0
Run Queue Flags: SUSPENDED | NONEMPTY
Current Process:
=dirty_cpu_scheduler:3
Scheduler Sleep Info Flags: SLEEPING | TSE_SLEEPING | WAITING
Scheduler Sleep Info Aux Work:
Current Process:
=dirty_cpu_scheduler:4
Scheduler Sleep Info Flags: SLEEPING | TSE_SLEEPING | WAITING | SUSPENDED
Scheduler Sleep Info Aux Work:
Current Process:
=dirty_cpu_run_queue
Run Queue Max Length: 0
Run Queue High Length: 0
Run Queue Normal Length: 0
Run Queue Low Length: 0
Run Queue Port Length: 0
Run Queue Flags: OUT_OF_WORK | HALFTIME_OUT_OF_WORK
=dirty_io_scheduler:5
Scheduler Sleep Info Flags: SLEEPING | TSE_SLEEPING | WAITING
Scheduler Sleep Info Aux Work:
Current Process:
=dirty_io_scheduler:6
Scheduler Sleep Info Flags: SLEEPING | TSE_SLEEPING | WAITING
Scheduler Sleep Info Aux Work:
Current Process:
=dirty_io_scheduler:7
Scheduler Sleep Info Flags: SLEEPING | TSE_SLEEPING | WAITING
Scheduler Sleep Info Aux Work:
```

**Operating environment**

- TeslaMate Version: v1.16.0
- Type of installation: Manual
- OS: [e.g. Ubuntu] Ubuntu 19.04

**Comment by ngardiner** [comment]
A side-effect of the crashes is that a lot of my charging data is now broken into smaller sessions without being recorded correctly, for example:

![image](https://user-images.githubusercontent.com/25845228/75332378-6215d680-58d8-11ea-9900-7e727517d816.png)

**Comment by adriankumpf** [comment]
Hmm this is a first.

It could be a problem with the Erlang/OTP installation. Are there any other error messages in the TeslaMate application logs?

**Comment by ngardiner** [comment]
No error, and no indication of any commonality before it happens, except that my observation is that it is more likely to happen when something (a drive or charge) is underway.

I've decided to do an upgrade to 20.04 LTS now that it is released, as the release of 20.04 officially makes 19.04 unsupported. I'll recompile after this and see if it improves stability.

I'll close this for now, however I do think the point about unfinished charge sessions may intersect with the other logged issue around the same topic - because TeslaMate's architecture is that it tracks the session end to end, if there's any interruption it does leave us with blank charging sessions. My preference is to preserve the data so if the official fix is to delete the session that's fine, but I would still manually update the charging_process entry as there should be enough charge data to populate most of the fields.

**Comment by adriankumpf** [comment]
> No error, and no indication of any commonality before it happens, except that my observation is that it is more likely to happen when something (a drive or charge) is underway.

That's very strange. Let us know if the same error occurs again with the new installation.

> My preference is to preserve the data so if the official fix is to delete the session that's fine, but I would still manually update the charging_process entry as there should be enough charge data to populate most of the fields.

I agree. Deletion may not always be the right choice.  It must be decided on a case-by-case basis. The updated documents describe how to finish an interrupted session: https://teslamate.readthedocs.io/en/latest/maintenance/manually_fixing_data.html
